relay-aws-extension/src/aws_extension.rs

```diff
@@ -0,0 +1,193 @@
+use actix::actors::signal::{Signal, SignalType};
+use actix::prelude::*;
+use failure::Fail;
+use relay_system::Controller;
+use reqwest::blocking::Client;
+use reqwest::StatusCode;
+use serde::Deserialize;
+use std::collections::HashMap;
+
+const EXTENSION_NAME: &str = "sentry-lambda-extension";
+const EXTENSION_NAME_HEADER: &str = "Lambda-Extension-Name";
+const EXTENSION_ID_HEADER: &str = "Lambda-Extension-Identifier";
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+struct RegisterResponse {
+    pub function_name: String,
+    pub function_version: String,
+    pub handler: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+struct Tracing {
+    pub r#type: String,
+    pub value: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+#[serde(rename_all = "UPPERCASE", tag = "eventType")]
+enum NextEventResponse {
+    #[serde(rename_all = "camelCase")]
+    Invoke {
+        deadline_ms: u64,
+        request_id: String,
+        invoked_function_arn: String,
+        tracing: Tracing,
+    },
+    #[serde(rename_all = "camelCase")]
+    Shutdown {
+        shutdown_reason: String,
+        deadline_ms: u64,
+    },
+}
+
+#[derive(Debug, Fail)]
+#[fail(display = "aws extension error")]
+/// Generic error in an AWS extension context
+pub struct AwsExtensionError;
+
+/// Actor implementing an AWS extension
+///
+/// Spawns an actor that
+/// * registers with the AWS extensions API
+/// * sends blocking NextEvent calls to the extensions API to get the next invocation
+///
+/// Each finished invocation immediately polls for the next event. Note that AWS might
+/// freeze the container indefinitely for unused lambdas and this request will also wait
+/// till things are active again.
+pub struct AwsExtension {
+    /// The base url for the AWS Extensions API
+    base_url: String,
+    /// The reqwest client needs to be blocking and with 0 timeout
+    /// because the container might get frozen if lambda is unused
+    reqwest_client: Client,
+    /// The extension id that will be retrieved on register and used
+    /// for subsequent requests
+    extension_id: Option<String>,
+}
+
+impl AwsExtension {
+    /// Creates a new `AwsExtension` instance.
+    pub fn new(aws_runtime_api: String) -> Result<Self, AwsExtensionError> {
+        let base_url = format!("http://{}/2020-01-01/extension", aws_runtime_api);
```

Comment:
Which additional validation do we need to put here to ensure this doesn't become an attack vector?